python/ray/experimental/data/dataset.py

```diff
@@ -106,7 +105,49 @@ def map_batches(self,
             ray_remote_args: Additional resource requirements to request from
                 ray (e.g., num_gpus=1 to request GPUs for the map tasks).
         """
-        raise NotImplementedError  # P0
+
+        def transform_builder(
+                build_block: Callable[[Block[T]], Block[U]]
+        ) -> Callable[[Iterable[Block[T]]], Iterable[ObjectRef[Block[U]]]]:
+            def transform(*blocks: Iterable[Block[T]]
+                          ) -> Iterable[ObjectRef[Block[U]]]:
+                # Need to ray.put so that the caller side
+                # can unflatten it without additional overhead.
+                return [ray.put(build_block(block)) for block in blocks]
+
+            return transform
+
+        def build_pandas_block(block: Block[T]) -> Block[U]:
+            builder = DelegatingArrowBlockBuilder()
+            df = block.to_pandas()
+            builder.add(fn(df))
+            return builder.build()
+
+        def build_arrow_block(block: Block[T]) -> Block[U]:
+            builder = DelegatingArrowBlockBuilder()
+            for row in block.iter_rows():
+                builder.add(fn(row))
+            return builder.build()
+
+        def get_batch_transform(batch_format: str) -> Any:
+            if batch_format == "pandas":
+                return transform_builder(build_pandas_block)
+            elif batch_format == "arrow":
+                return transform_builder(build_arrow_block)
+            else:
+                raise ValueError(
+                    f"The given batch format: {batch_format} "
+                    f"is not supported. Supported types: {BatchType}")
+
+        transform = get_batch_transform(batch_format)
+        compute = get_compute(compute)
+        batch_size = 1 if batch_size is None else batch_size
+        if batch_size < 1:
+            raise ValueError("Batch size cannot be negative or 0")
+
+        return Dataset(
+            compute.apply_batch(transform, ray_remote_args, self._blocks,
```

Comment:
I don't think we need apply_batch. We can just say the batch size "max batch size". So if a block only has 3 elements in it, we just pass a batch of 3 elems.

So this way we can use normal .apply() and the batching happens in the per-block transform function noly.